x-pack/plugins/security_solution/server/lib/entity_analytics/asset_criticality/asset_criticality_data_client.ts

```diff
@@ -135,6 +133,149 @@ export class AssetCriticalityDataClient {
     return doc;
   }
 
+  public async bulkUpsert(
+    records: AssetCriticalityUpsert[]
+  ): Promise<
+    Array<
+      | { id: string; error: string }
+      | { record: AssetCriticalityRecord; result: 'created' | 'updated' }
+    >
+  > {
+    if (records.length === 0) {
+      return [];
+    }
+
+    const docs = records.map((record) => ({
+      _id: createId(record),
+      doc: {
+        id_field: record.idField,
+        id_value: record.idValue,
+        criticality_level: record.criticalityLevel,
+        '@timestamp': new Date().toISOString(),
+      },
+    }));
+
+    const body = docs.flatMap(({ doc, _id }) => [
+      { update: { _id } },
+      { doc, doc_as_upsert: true },
+    ]);
+
+    const response = await this.options.esClient.bulk({
+      body,
+      index: this.getIndex(),
+    });
+
+    return response.items.map((item, index) => {
+      const doc = docs[index];
+      if (item.update && item.update.error) {
+        return { id: doc._id, error: item.update.error.reason || item.update.error.type };
+      } else {
+        if (item?.update?.result === 'created') {
+          return { record: doc.doc, result: 'created' };
+        } else {
+          return { record: doc.doc, result: 'updated' };
+        }
+      }
+    });
+  }
+
+  /**
+   * Bulk upsert asset criticality records from a stream.
+   * @param recordsStream a stream of records to upsert, records may also be an error e.g if there was an error parsing
+   * @param batchSize the number of records to upsert in a single batch
+   * @returns an object containing the number of records updated, created, errored, and the total number of records processed
+   * @throws an error if the stream emits an error
+   * @remarks
+   * - The stream must emit records in the format of {@link AssetCriticalityUpsert} or an error instance
+   * - The stream must emit records in the order they should be upserted
+   * - The stream must emit records in a valid JSON format
+   * - We allow errors to be emitted in the stream to allow for partial upserts and to maintain the order of records
+   **/
+  public bulkUpsertFromStream = async ({
+    recordsStream,
+    batchSize,
+  }: {
+    recordsStream: NodeJS.ReadableStream;
+    batchSize: number;
+  }): Promise<AssetCriticalityCsvUploadResponse> => {
+    const errors: AssetCriticalityCsvUploadResponse['errors'] = [];
+    const stats: AssetCriticalityCsvUploadResponse['stats'] = {
+      updated: 0,
+      created: 0,
+      errors: 0,
+      total: 0,
+    };
+
+    async function* recordsGenerator(): AsyncGenerator<
```

Comment:
Why do we need this `AsyncGenerator`?
It seems like we can just iterate directly over the records stream?

Better yet, I'd suggest using the event handler approach via the `'data` and `'end'` events instead of `for await` loops. 
In general, I think making async code look synchronous can lead to confusing code, and just sticking to the basic, simple Stream interface is probably better

Reply:
I refactored to a generator in https://github.com/elastic/kibana/pull/179930/commits/faab3922b8204cda635ba0117c7f73a596b3857a based on roberts suggestion. I can see advantages of both approaches, I do think the generator makes the batching code neater, but there isnt a big difference to me. I think we should discuss this in Roberts code review meeting later